akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala

```diff
@@ -941,35 +940,69 @@ private[cluster] class ClusterCoreDaemon(publisher: ActorRef) extends Actor with
   /**
    * Initiates a new round of gossip.
    */
-  def gossip(): Unit = {
-
+  def gossip(): Unit =
     if (!isSingletonCluster) {
-      val localGossip = latestGossip
+      val gossipTargets =
+        // TODO break this two methods out and unit test?
+        if (latestGossip.isMultiDc) multiDcGossipTargets()
+        else localDcGossipTargets()
+
+      selectRandomNode(gossipTargets) match {
+        case Some(peer) ⇒ gossipTo(peer)
+        case None       ⇒ // nothing to see here
+      }
+    }
 
-      val preferredGossipTargets: Vector[UniqueAddress] =
-        if (ThreadLocalRandom.current.nextDouble() < adjustedGossipDifferentViewProbability) {
-          // If it's time to try to gossip to some nodes with a different view
-          // gossip to a random alive member with preference to a member with older gossip version
-          localGossip.members.collect {
-            case m if !localGossip.seenByNode(m.uniqueAddress) && validNodeForGossip(m.uniqueAddress) ⇒
-              m.uniqueAddress
-          }(breakOut)
-        } else Vector.empty
-
-      if (preferredGossipTargets.nonEmpty) {
-        val peer = selectRandomNode(preferredGossipTargets)
-        // send full gossip because it has different view
-        peer foreach gossipTo
-      } else {
-        // Fall back to localGossip; important to not accidentally use `map` of the SortedSet, since the original order is not preserved)
-        val peer = selectRandomNode(localGossip.members.toIndexedSeq.collect {
-          case m if validNodeForGossip(m.uniqueAddress) ⇒ m.uniqueAddress
-        })
-        peer foreach { node ⇒
-          if (localGossip.seenByNode(node)) gossipStatusTo(node)
```

Comment:
In the current logic we mix selection and side effect, which I was trying to get away from. I moved this all-nodes-has-seen-gossip fallback inside of the logic that gives back the set of nodes to choose randomly among. In that case the side-effect logic doing the actual gossip doesn't know about which list of nodes that was chosen, so I wondered if after selecting that random node, it would hurt to too the seenByNode check and if seen only gossip status. I think it should give the same logic as currently for local (and do that same thing for cross-dc gossip)